src/test/java/com/aws/greengrass/clientdevices/auth/infra/NetworkStateTest.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.clientdevices.auth.infra;
+
+import com.aws.greengrass.mqttclient.CallbackEventManager;
+import com.aws.greengrass.mqttclient.MqttClient;
+import com.aws.greengrass.testcommons.testutilities.GGExtension;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import software.amazon.awssdk.crt.mqtt.MqttClientConnectionEvents;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.not;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.verify;
+
+
+@ExtendWith({MockitoExtension.class, GGExtension.class})
+public class NetworkStateTest {
+    @Mock
+    MqttClient mqttClient;
+    @Captor
+    ArgumentCaptor<CallbackEventManager.OnConnectCallback> onConnectCaptor;
+    @Captor
+    ArgumentCaptor<MqttClientConnectionEvents> connectionEventsArgumentCaptor;
+    private ForkJoinPool fjp = new ForkJoinPool();
+    private NetworkState networkState;
+
+    @BeforeEach
+    void beforeEach() {
+        fjp = new ForkJoinPool();
+        networkState = new NetworkState(mqttClient, fjp);
+        verify(mqttClient).addToCallbackEvents(onConnectCaptor.capture(), connectionEventsArgumentCaptor.capture());
+    }
+
+    @AfterEach
+    void afterEach() {
+        fjp.shutdown();
+    }
+
+    void assertNetworkDomainEventOnNetworkEvent(Runnable networkEvent, NetworkState.ConnectionState expectedState)
+            throws InterruptedException {
+        CountDownLatch cdl = new CountDownLatch(1);
```

Comment:
I like this helper but it hides what the steps of the tests are. When we read the test we just see a call to an assertion  without any step that setup pre conditions, run the system under test and then assert on things. I think we might want to split it into


```java
private Runnable givenSubscribreToNetworkChanges() {
          CountDownLatch cdl = new CountDownLatch(1);


        Consumer<NetworkState.ConnectionState> networkStateConsumer = (state) -> {
            connState.set(state);
            callbackThreadId.set(Thread.currentThread().getId());
            cdl.countDown();
        };
        networkState.registerHandler(networkStateConsumer);

      return () -> {
             networkEvent.run();
            cdl.await(1, TimeUnit.SECONDS);
      }
}

```

Then your test becomes

```java
 // Given
  Runnable onNetworkUp = givenSubscribreToNetworkChanges(() -> onConnectCaptor.getValue().onConnect(true),
                NetworkState.ConnectionState.NETWORK_UP);

// When 
         onNetworkUp.run();

// Then
        assertThat(callbackThreadId.get(), is(not(curThreadId)));
        assertThat(connState.get(), is(NetworkState.ConnectionState.NETWORK_UP));
```

Reply:
The snippet was meant as an example I did not run it.  The `@ParametrizedTest` looks good as well it is achieving something similar, the only thing I would change if the method name to something like `GIVEN_subscriptionToNetworkChanges_WHEN_connectionStateChanges_THEN_subscribersRunOnASeparateThread`